examples/chef/common/chef-rvc-mode-delegate.cpp

```diff
@@ -0,0 +1,210 @@
+    {
+        response.status = to_underlying(ModeBase::StatusCode::kGenericFailure);
+        response.statusText.SetValue(chip::CharSpan::fromCharString("Change to the mapping mode is only allowed from idle"));
+        return;
+    }
+
+    response.status = to_underlying(ModeBase::StatusCode::kSuccess);
+}
+
+CHIP_ERROR RvcRunModeDelegate::GetModeLabelByIndex(uint8_t modeIndex, chip::MutableCharSpan & label)
+{
+    if (modeIndex >= ArraySize(kModeOptions))
+    {
+        return CHIP_ERROR_PROVIDER_LIST_EXHAUSTED;
+    }
+    return chip::CopyCharSpanToMutableCharSpan(kModeOptions[modeIndex].label, label);
+}
+
+CHIP_ERROR RvcRunModeDelegate::GetModeValueByIndex(uint8_t modeIndex, uint8_t & value)
+{
+    if (modeIndex >= ArraySize(kModeOptions))
+    {
+        return CHIP_ERROR_PROVIDER_LIST_EXHAUSTED;
+    }
+    value = kModeOptions[modeIndex].mode;
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR RvcRunModeDelegate::GetModeTagsByIndex(uint8_t modeIndex, List<ModeTagStructType> & tags)
+{
+    if (modeIndex >= ArraySize(kModeOptions))
+    {
+        return CHIP_ERROR_PROVIDER_LIST_EXHAUSTED;
+    }
+
+    if (tags.size() < kModeOptions[modeIndex].modeTags.size())
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+
+    std::copy(kModeOptions[modeIndex].modeTags.begin(), kModeOptions[modeIndex].modeTags.end(), tags.begin());
+    tags.reduce_size(kModeOptions[modeIndex].modeTags.size());
+
+    return CHIP_NO_ERROR;
+}
+
+ModeBase::Instance * RvcRunMode::Instance()
+{
+    return gRvcRunModeInstance;
+}
+
+void RvcRunMode::Shutdown()
+{
+    if (gRvcRunModeInstance != nullptr)
+    {
+        delete gRvcRunModeInstance;
+        gRvcRunModeInstance = nullptr;
+    }
+    if (gRvcRunModeDelegate != nullptr)
+    {
+        delete gRvcRunModeDelegate;
+        gRvcRunModeDelegate = nullptr;
+    }
+}
+
+void emberAfRvcRunModeClusterInitCallback(chip::EndpointId endpointId)
+{
+    VerifyOrDie(endpointId == 1); // this cluster is only enabled for endpoint 1.
+    VerifyOrDie(gRvcRunModeDelegate == nullptr && gRvcRunModeInstance == nullptr);
+    gRvcRunModeDelegate = new RvcRunMode::RvcRunModeDelegate;
+    gRvcRunModeInstance =
+        new ModeBase::Instance(gRvcRunModeDelegate, 0x1, RvcRunMode::Id, chip::to_underlying(RvcRunMode::Feature::kOnOff));
+    gRvcRunModeInstance->Init();
+}
+#endif // ZCL_USING_RVC_RUN_MODE_CLUSTER_SERVER
+
+#ifdef ZCL_USING_RVC_CLEAN_MODE_CLUSTER_SERVER
+#include <chef-rvc-mode-delegate.h>
+using namespace chip::app::Clusters::RvcCleanMode;
+static RvcCleanModeDelegate * gRvcCleanModeDelegate = nullptr;
+static ModeBase::Instance * gRvcCleanModeInstance   = nullptr;
+
+CHIP_ERROR RvcCleanModeDelegate::Init()
+{
+    return CHIP_NO_ERROR;
+}
+
+void RvcCleanModeDelegate::HandleChangeToMode(uint8_t NewMode, ModeBase::Commands::ChangeToModeResponse::Type & response)
+{
+    uint8_t rvcRunCurrentMode = gRvcRunModeInstance->GetCurrentMode();
```

Comment:
Technically `gRvcRunModeInstance` is in a `#ifdef ZCL_USING_RVC_RUN_MODE_CLUSTER_SERVER` and not `ZCL_USING_RVC_CLEAN_MODE_CLUSTER_SERVER`. Sure you will get a compiler error if you don't have this configured properly. If this is supposed to be this way can you include a comment on why this is done this way?

Reply:
Said in a more actionable way, I think it make sense that you have a `static_assert` here if `ZCL_USING_RVC_RUN_MODE_CLUSTER_SERVER` is not defined with a error saying that it is expected that if clean mode cluster server is included this app expected the run mode cluster to also be present.

Reply:
Thanks for the correction. After checking spec: `RVC Run Mode` is Mandatory  while `RVC Clean Mode` is Optional, I decided to write codes as below. WDYT?

```
#ifdef ZCL_USING_RVC_RUN_MODE_CLUSTER_SERVER
// .. (rvc run codes) ...
#ifdef ZCL_USING_RVC_CLEAN_MODE_CLUSTER_SERVER
// .. (rvc clean codes) ...
#endif // ZCL_USING_CLEAN_RUN_MODE_CLUSTER_SERVER
#endif // ZCL_USING_RVC_RUN_MODE_CLUSTER_SERVER
```

Reply:
LGTM